### PR TITLE
Implement push-docker entrypoint

### PR DESCRIPTION
**Pull request by midnightercz**
Push-docker entrypoint includes all steps needed to push container
images to quay. Entrypoint is implemented as python module only
so it has to be used within pub only so far

### CODE REVIEW

File: pubtools/_quay/utils/stepper.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+
+def isodate_now():
+    """Return current datetime in iso8601 format."""
+    return "%s%s" % (datetime.datetime.utcnow().isoformat(), "Z")
+
+
+class StepFailedError(Exception):
+    """Exception indidating failure of a step."""
+
+    pass
```
Comment (querti) [inline pubtools/_quay/utils/stepper.py:14]: I think this is not needed.

File: pubtools/_quay/utils/stepper.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+
+    def __init__(
+        self, uid, step_args, step_kwargs, shared_results, external_resources=None
+    ):
+        """Initilize the step.
+
+        Args:
+            uid: (str)
+                An unique id for indentifying two steps of the same class
+            uid: (step_args)
```
Comment (querti) [inline pubtools/_quay/utils/stepper.py:117]: wrong name and type

File: pubtools/_quay/utils/stepper.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+                Arguments used for the step
+            uid: (step_kwargs)
```
Comment (querti) [inline pubtools/_quay/utils/stepper.py:119]: wrong name and type

File: pubtools/_quay/utils/stepper.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+                Dictionary of a name arguments for the step
+            shared_results: (dict-like object)
+                Object to store shared data between steps
+            external_resources: any-object
+                Reference for external resources (in any form) which are constant and
+                shouldn't be part of the step state or step data
+        """
+        self.step_args = step_args
```
Comment (querti) [inline pubtools/_quay/utils/stepper.py:127]: this is unnecessary right?

File: pubtools/_quay/utils/stepper.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+            arg.value if isinstance(arg, Secret) else arg for arg in step_args
+        ]
+        self.step_kwargs = dict(
+            [
+                (k, arg.value) if isinstance(arg, Secret) else (k, arg)
+                for k, arg in step_kwargs.items()
+            ]
+        )
+
+        self.step_kwargs = step_kwargs
```
Comment (querti) [inline pubtools/_quay/utils/stepper.py:149]: you are overwriting the previous step_kwargs definition

File: pubtools/_quay/utils/stepper.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+        self.results = StepResults()
+
+    @property
+    def fullname(self):
+        """Full name of class instance."""
+        return "%s:%s" % (self.NAME, self.uid)
+
+    @property
+    def state(self):
+        """Property for retreiving class state."""
```
Comment (querti) [inline pubtools/_quay/utils/stepper.py:169]: retrieving

File: pubtools/_quay/utils/stepper.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+        self.stats["finished"] = isodate_now()
+        self.stats["skipped"] = self.skip
+        self.stats["skip"] = self.skip
+        self.stats["skip_reason"] = self.skip_reason
+        self.stats["state"] = self.state
+
+    def _store_results(self):
+        self._shared_results[self.fullname] = self.results
+
+    @abc.abstractmethod
```
Comment (querti) [inline pubtools/_quay/utils/stepper.py:243]: Is this Python 2.6 compatible?

File: pubtools/_quay/quay_steps.py [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+        "log_debug": log debug callback or None
+        "log_info": log info callback or None
+        "log_warning": log warning callback or None
+        "log_error": log error callback or None
+        "hub": pubhub xmlrpcserver instance
+        "task_id": task id
+        "target_name": target name
+    Produced details:
+        {
+            "items": { <repo>: { <tag> { "state": ready|error|pushed, "source": <pull_url>}}}
```
Comment (querti) [inline pubtools/_quay/quay_steps.py:408]: This is incorrect isn't it?

File: pubtools/_quay/quay_steps.py [hunk 9]
```diff
@@ -0,0 +1,631 @@
...
+        entry_point = (
+            "pubtools-pyxis",
+            "console_scripts",
+            "pubtools-pyxis-get-operator-indices",
+        )
+        entry_point_fn = pkg_resources.load_entry_point(*entry_point)
+        args = [
+            "cmd",
+            "--pyxis-server",
+            self.step_kwargs["pyxis_server"],
```
Comment (querti) [inline pubtools/_quay/quay_steps.py:475]: This stuff is in target settings right? push_docker function doesn't produce this data

**Review by querti — CHANGES_REQUESTED** [review]

File: pubtools/_quay/push_docker.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+
+from .utils.stepper import Stepper
+from .utils.logger import Logger
+
+import pushcollector
+
+CLI_DESCRIPTION = ""
+
+
+def log_push_items(signing_key, items=None):
```
[the remark this answers is not in the document]
Comment (midnightercz) [reply pubtools/_quay/push_docker.py:24]: that's original name of the method taken from pub. It basically update push items details and save them in hub

File: pubtools/_quay/utils/stepper.py (hunk 5 above)
Comment (midnightercz) [reply pubtools/_quay/utils/stepper.py:149]: ah, right, I forgot to remove it

File: pubtools/_quay/utils/stepper.py (hunk 7 above)
Comment (midnightercz) [reply pubtools/_quay/utils/stepper.py:243]: I think it is https://docs.python.org/2.6/library/abc.html#module-abc

File: pubtools/_quay/quay_steps.py (hunk 9 above)
Comment (midnightercz) [reply pubtools/_quay/quay_steps.py:476]: fixed

File: pubtools/_quay/quay_steps.py [hunk 11]
```diff
@@ -205,6 +205,45 @@ def _run(self, on_update=None):
         self.results.results = operator_push_items
 
 
+class StepSanitizeRepositories(Step):
+    """Filter out push items which are not operators and store result.
```
Comment (querti) [inline pubtools/_quay/quay_steps.py:209]: Is this correct?

**Review by querti — CHANGES_REQUESTED** [review]

File: pubtools/_quay/push_docker.py [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+            shared_data,
+            external_resources=common_external_res,
+        )
+    )
+    stepper.add_step(
+        StepPushOperators(
+            "1",
+            ("StepSanitizeOperatorPushItems:1", target_settings),
+            {
+                "docker_reference_registry": target_settings.get(
```
Comment (emilyzheng) [inline pubtools/_quay/push_docker.py:145]: docker_reference_registry is in target_settings["docker_settings"].

File: pubtools/_quay/quay_steps.py [hunk 13]
```diff
@@ -0,0 +1,671 @@
...
+        log_error = self.external_resources.get("log_error", lambda *args, **kwargs: ())
+
+        container_push_items = []
+        _on_update = on_update if on_update else lambda: ()
+        failed = False
+        for i, item in enumerate(push_items):
+            log_info("Processing %s", str(item))
+            item_error = False
+            item_state = "ok"
+            if item.file_type != "docker_v2":
```
Comment (emilyzheng) [inline pubtools/_quay/quay_steps.py:58]: File type "docker_v2" was removed from Pub. Use "docker" here?

File: pubtools/_quay/quay_steps.py (hunk 11 above)
Comment (midnightercz) [reply pubtools/_quay/quay_steps.py:209]: ah, copy paste issue

File: pubtools/_quay/push_docker.py (hunk 12 above)
Comment (midnightercz) [reply pubtools/_quay/push_docker.py:145]: Yes, that's for pulp targets, but I think for quay it doesn't make sense separate this in extra dictionary.

File: pubtools/_quay/quay_steps.py (hunk 13 above)
Comment (midnightercz) [reply pubtools/_quay/quay_steps.py:58]: right, I forgot it was changed. Thanks for pointing this out

File: pubtools/_quay/push_docker.py (hunk 12 above)
Comment (emilyzheng) [reply pubtools/_quay/push_docker.py:145]: How about docker_container_signing_enabled?

Comment (midnightercz) [reply pubtools/_quay/push_docker.py:145]: good point, I'll change it to container_signing_enabled

**Review by querti — APPROVED** [review]

File: pubtools/_quay/push_docker.py (hunk 12 above)
Comment (emilyzheng) [reply pubtools/_quay/push_docker.py:145]: But why it needs to be in target_settings["docker_settings"] not target_settings? In Line 123.

**Review by emilyzheng — APPROVED** [review]

**Pull request merged**